### PR TITLE
Refactor ExecParams for Deploy and Update action

### PR DESCRIPTION
**Pull request by vuvoth**
## Proposed Changes

+ Update web/types.
+ appName is required. 
+ Update {runtime.algob}.deployApp().
+ Update {runtime,algob}.updateApp().
+ Remove deployer.getAppByFile.
+ Remove runtime.addApp();
+ Update tests. 

## TODO: 
- [x] update UpdateApp method. 
- [x] Update docs. 
- [x] Update Changelog
- [x] Update web-mode.

### CODE REVIEW

File: packages/web/src/types.ts [hunk 1]
```diff
@@ -169,20 +169,45 @@ export type DeployASAParam = BasicParams & {
...
+export type SourceFile = {
+	metaType: MetaType.FILE;
+	approvalProgramPath: string;
+	clearProgramPath: string;
+};
+
+export type SourceString = {
+	metaType: MetaType.STRING;
+	approvalProgramSource: string;
+	clearProgramSource: string;
```
Comment (robert-zaremba) [inline packages/web/src/types.ts:187]:
```suggestion
	approvalProgramCode: string;
	clearProgramCode: string;
```

File: packages/web/src/types.ts [hunk 2]
```diff
@@ -169,20 +169,45 @@ export type DeployASAParam = BasicParams & {
...
+};
+
+export type SourceBytes = {
+	metaType: MetaType.BYTES;
+	approvalProgram: Uint8Array;
+	clearProgram: Uint8Array;
```
Comment (robert-zaremba) [inline packages/web/src/types.ts:193]:
```suggestion
	approvalProgramBytes: Uint8Array;
	clearProgramBytes: Uint8Array;
```

File: packages/web/src/types.ts [hunk 3]
```diff
@@ -169,20 +169,45 @@ export type DeployASAParam = BasicParams & {
...
+	clearProgramPath: string;
+};
+
+export type SourceString = {
+	metaType: MetaType.STRING;
+	approvalProgramSource: string;
+	clearProgramSource: string;
+};
+
+export type SourceBytes = {
```
Comment (robert-zaremba) [inline packages/web/src/types.ts:190]: Is this compiled program? If yes, then let's rename to SourceCompiled

File: packages/web/src/types.ts [hunk 4]
```diff
@@ -169,20 +169,45 @@ export type DeployASAParam = BasicParams & {
...
+	metaType: MetaType.BYTES;
+	approvalProgram: Uint8Array;
+	clearProgram: Uint8Array;
+};
+
+export type SourceContract = SourceFile | SourceString | SourceBytes;
```
Comment (robert-zaremba) [inline packages/web/src/types.ts:196]:
```suggestion
export type SmartContract = SourceFile | SourceString | SourceBytes;
```

File: packages/web/src/types.ts [hunk 5]
```diff
@@ -169,20 +169,45 @@ export type DeployASAParam = BasicParams & {
...
+	BYTES,
+}
+
+export type SourceFile = {
+	metaType: MetaType.FILE;
+	approvalProgramPath: string;
+	clearProgramPath: string;
+};
+
+export type SourceString = {
```
Comment (robert-zaremba) [inline packages/web/src/types.ts:184]:
```suggestion
export type SourceCode = {
```

**Review by robert-zaremba — COMMENTED** [review]
lgtm, added few suggestions to rename types.

File: examples/asa/scripts/3-contract-owned-asa.js [hunk 6]
```diff
@@ -21,18 +21,18 @@ async function run(runtimeEnv, deployer) {
...
 	const appInfo = await deployer.deployApp(
-		"5-contract-asa-stateful.py", // approval program
-		"5-clear.py", // clear program
+		alice,
 		{
-			sender: alice,
+			appName: "StatefulASA_App",
+			metaType: types.MetaType.FILE,
+			approvalProgramFileName: "5-contract-asa-stateful.py", // approval program
+			clearProgramFileName: "5-clear.py", // clear program
```
Comment (robert-zaremba) [inline examples/asa/scripts/3-contract-owned-asa.js:29]:
```suggestion
			clearProgramFilename: "5-clear.py", // clear program
```

**Review by robert-zaremba — COMMENTED** [review]
added small rename suggestion for the type

File: examples/bond/test/failing-tests.js [hunk 7]
```diff
@@ -89,7 +88,7 @@ describe("Bond token failing tests", function () {
 			creator: { ...bondTokenCreator.account, name: "bond-token-creator" },
 		}).assetIndex;
 
-		const creationFlags = Object.assign({}, flags);
+		const appDef = Object.assign({}, appStorageConfig);
```
Comment (sczembor) [inline examples/bond/test/failing-tests.js:91]:
```suggestion
		const appDefinition = Object.assign({}, appStorageConfig);
```

I think a longer a bit more explicit names are easier to read

File: examples/crowdfunding/test/crowdfundingTest.js [hunk 8]
```diff
@@ -91,13 +97,12 @@ describe("Crowdfunding Tests", function () {
 		 * Note: - In this example timestamps are commented because it is possible
 		 * that network timestamp and system timestamp may not be in sync.
 		 */
-		const creationFlags = Object.assign({}, flags);
+		const appDef = Object.assign({}, appDefinition);
```
Comment (sczembor) [inline examples/crowdfunding/test/crowdfundingTest.js:100]: same as above

File: examples/crowdfunding/test/crowdfundingTest.js [hunk 9]
```diff
@@ -41,8 +44,11 @@ describe("Crowdfunding Tests", function () {
 		donor = new AccountStore(initialDonorBalance);
 		runtime = new Runtime([master, creator, escrow, donor]);
 
-		flags = {
-			sender: creator.account,
+		appDefinition = {
+			appName: "crowdFundingApp",
```
Comment (sczembor) [inline examples/crowdfunding/test/crowdfundingTest.js:48]:
```suggestion
			appName: "crowdfundingApp",
```

File: examples/crowdfunding/test/crowdfundingTest.js [hunk 10]
```diff
@@ -41,8 +44,11 @@ describe("Crowdfunding Tests", function () {
...
+			metaType: types.MetaType.FILE,
+			approvalProgramFilename: crowdFundApprovalFileName,
```
Comment (sczembor) [inline examples/crowdfunding/test/crowdfundingTest.js:50]:
```suggestion
			approvalProgramFilename: crowdfundApprovalFilename,
```

File: examples/crowdfunding/test/failing-tests.js [hunk 11]
```diff
@@ -55,8 +55,11 @@ describe("Crowdfunding Test - Failing Scenarios", function () {
 			convert.addressToPk(creator.address),
 			convert.uint64ToBigEndian(fundCloseDate.getTime()),
 		];
-		const creationFlags = {
-			sender: creator.account,
+		const appDef = {
+			appName: "crowdFundingApp",
```
Comment (sczembor) [inline examples/crowdfunding/test/failing-tests.js:59]:
```suggestion
			appName: "crowdfundingApp",
```

File: examples/crowdfunding/test/happypaths.js [hunk 12]
```diff
@@ -67,7 +70,13 @@ describe("Crowdfunding Tests - Happy Paths", function () {
 		runtime = new Runtime([master, creator, donor, fundReceiver]);
 
 		applicationId = 1;
-		creator.addApp(applicationId, creationFlags, approvalProgram, clearProgram);
+		// addApp only work with AppDefinition from Source
```
Comment (sczembor) [inline examples/crowdfunding/test/happypaths.js:73]:
```suggestion
		// addApp only works with AppDefinition from Source
```

File: packages/algob/src/lib/algo-operator.ts [hunk 13]
```diff
@@ -577,9 +579,47 @@ export class AlgoOperatorImpl implements AlgoOperator {
...
-		return await this.compileOp.ensureCompiled(name, force, scTmplParams);
+		return await this.compileOp.ensureCompiled(name, source, force, scTmplParams);
+	}
+
+	async compileApplication(
+		appName: string,
+		source: wtypes.SmartContract,
+		scTmplParams?: SCParams
+	): Promise<wtypes.SourceBytes> {
+		// we don't need compile bytes source.
```
Comment (sczembor) [inline packages/algob/src/lib/algo-operator.ts:594]:
```suggestion
		// in case of bytes source we do not need to compile it
```

**Review by sczembor — APPROVED** [review]
lgtm! left few comments

File: CHANGELOG.md [hunk 14]
```diff
@@ -71,6 +71,145 @@ console.log("txn1 information: ", receipts[2]);
...
+- We add `SmartContract` types for describe the application source code. Details: 
+```ts
+  // from file
+  type SourceFile = {
+    metaType: MetaType.FILE;
+    approvalProgramFilename: string;
+    clearProgramFilename: string;
+  };
+
+  // from raw teal source code.
```
Comment (robert-zaremba) [inline CHANGELOG.md:83]:
```suggestion
  // from teal source code (string).
```

File: CHANGELOG.md [hunk 15]
```diff
@@ -71,6 +71,145 @@ console.log("txn1 information: ", receipts[2]);
...
+  type SourceCode = {
+    metaType: MetaType.SOURCE_CODE;
+    approvalProgramCode: string;
+    clearProgramCode: string;
+  };
+
+  // from compiled source code. 
+  type SourceBytes = {
```
Comment (robert-zaremba) [inline CHANGELOG.md:91]:
```suggestion
  type SourceCompiled = {
```

File: CHANGELOG.md [hunk 16]
```diff
@@ -71,6 +71,145 @@ console.log("txn1 information: ", receipts[2]);
...
+
+- We change ``: Details(please check packages/web/src/types.ts):
+
+```ts
+
+export type AppDefinitionFromFile = StorageConfig & AppOptionalFlags & SourceFile;
+
+export type AppDefinitionFromSource = StorageConfig & AppOptionalFlags & SourceCode;
+
+export type AppDefinitionFromSourceCompiled = StorageConfig & AppOptionalFlags & SourceBytes;
```
Comment (robert-zaremba) [inline CHANGELOG.md:106]:
```suggestion
export type AppDefinitionFromSourceCompiled = StorageConfig & AppOptionalFlags & SourceCompiled;
```

File: packages/web/src/types.ts [hunk 17]
```diff
@@ -169,30 +169,62 @@ export type DeployASAParam = BasicParams & {
...
+	clearProgramFilename: string;
+};
+
+export type SourceCode = {
+	metaType: MetaType.SOURCE_CODE;
+	approvalProgramCode: string;
+	clearProgramCode: string;
+};
+
+export type SourceBytes = {
```
Comment (robert-zaremba) [inline packages/web/src/types.ts:198]:
```suggestion
Compiled bytes of a TEAL program.
export type SourceCompiled = {
```

**Review by robert-zaremba — APPROVED** [review]
Let's rename SourceBytes (which might be confusing) to SourceCompiled

**Pull request merged**